### DISCUSSION TITLE
can everyone explain the circular symlinks

### DISCUSSION THREAD

**Discussion by smallwebbird** [pnpm/pnpm#4207] [Q&A] [answered]
When I read this article [https://pnpm.io/blog/2020/05/27/flat-node-modules-is-not-the-only-way](https://pnpm.io/blog/2020/05/27/flat-node-modules-is-not-the-only-way)，I'm a little confused about the following paragraph.
`All the dependencies of express are symlinks to appropriate directories in node_modules/.pnpm/. Placing dependencies of express one level up allows avoiding circular symlinks.`, Can anyone explain it? thank you

**Comment by zkochan** [comment, marked as answer]
Let's say bar is a dep of foo and foo is a dep of foo. This is how the virtual store will look like:

```
.pnpm
  bar@1.0.0
    node_modules
      bar
      foo --> ../../foo@1.0.0/node_modules/foo
  foo@1.0.0
    node_modules
      foo
      bar --> ../../bar@1.0.0/node_modules/bar
```

The symlinks to foo and bar are not circular.

But we'd move the dependencies into a subfoler like this:

```
.pnpm
  bar@1.0.0
    node_modules
      bar
        node_modules
          foo --> ../../../../foo@1.0.0/node_modules/foo
  foo@1.0.0
    node_modules
      foo
        node_modules
          bar --> ../../../../bar@1.0.0/node_modules/bar
```

In this case, there are circular symlinks. The bar inside foo@1.0.0 is resolved to `bar@1.0.0/node_modules/bar`, then the foo inside `bar@1.0.0/node_modules/bar/node_modules/foo` is resolved to `foo@1.0.0/node_modules/foo`, which has a symlink to `bar@1.0.0/node_modules/bar` again in its subdirectories.

In VS code if you would try to open these directories, you'd get something like this:

```
.pnpm
  bar@1.0.0
    node_modules
      bar
        node_modules
          foo
            node_modules
              bar
                node_modules
                  foo
                    node_modules
                      bar
                        ....this goes indefinitely 
  foo@1.0.0
    node_modules
      foo
        node_modules
          bar
           node_modules
             foo
               node_modules
                 bar
                   ...this goes indefinitely
```

**Reply by huntye1** [reply]
I think both example can work when the code try to resolve the package. except the reason of VS code, is there any other reason to change the struct?

**Reply by zkochan** [reply]
for Node.js both work but some tools that scan node_modules will go into an infinite loop.
